include/openmc/static_map.h

```diff
@@ -0,0 +1,251 @@
+#ifndef OPENMC_STATIC_MAP_H
+#define OPENMC_STATIC_MAP_H
+
+#include <algorithm> // for stable_sort, find_if, lower_bound
+#include <utility> // for pair
+
+#include "openmc/vector.h"
+
+namespace openmc {
+
+//==========================================================================
+// Integer hash function
+//==========================================================================
+
+//! Hash function for integers that simply returns the integer.
+//
+//! If the values in the hash table are consecutive starting from zero, this
+//! will produce no collisions. Otherwise, all bets are off.
+//
+//! TODO: explore the performance of different hash functions
+template<typename Key>
+struct integer_hash
+{
+  using value_type = Key;
+  using result_type = std::size_t;
+
+  result_type operator()(value_type value) const
+  {
+    return static_cast<result_type>(value);
+  }
+};
+
+//==========================================================================
+// Class declarations
+//==========================================================================
+
+//! Fixed-size map with on-device access.
+//
+//! This is a read-only hash table that supports on-device lookups. The table
+//! is constructed on the host and must be "finalized" before it can be used to
+//! search for items. Adapted from the \c Static_Device_Map class in Oak Ridge
+//! National Laboratory's SCALE code system.
```

Comment:
```suggestion
//! National Laboratory's SCALE code system with help from 
//! Seth Johnson, Tom Evans, and Steven Hamilton.
```

Reply:
Might be nice to call them out in the comment here, but totally up to you if you think it's needed or not in this context.

Reply:
Yep I think that's a good call.